### PR TITLE
Disambiguate GenAPI calls to base constructors

### LINKED ISSUE #31655: GenAPI can generate ambiguous call to base constructor
When GenAPI defines arguments to a base constructor, it uses `default!` but this can lead to an ambiguous call when the base type has constructor overloads with the same number of parameters but different types.

An example of this is the generated code for NuGet.Protocol.6.5.0:

```c#
public partial class SourcePackageDependencyInfo : Packaging.Core.PackageDependencyInfo
{
    public SourcePackageDependencyInfo(Packaging.Core.PackageIdentity identity, System.Collections.Generic.IEnumerable<Packaging.Core.PackageDependency> dependencies, bool listed, SourceRepository source, System.Uri downloadUri, string packageHash) : base(default!, default!) { }
```

This leads to the following compiler error:

```
error CS0121: The call is ambiguous between the following methods or properties: 'PackageDependencyInfo.PackageDependencyInfo(PackageIdentity, IEnumerable<PackageDependency>)' and 'PackageDependencyInfo.PackageDependencyInfo(string, NuGetVersion)
```

This can be solved by explicitly casting `default` to the desired type:

```c#
: base((Packaging.Core.PackageIdentity)default!, (System.Collections.Generic.IEnumerable<Packaging.Core.PackageDependency>)default!) { }
```

cc @andriipatsula @ViktorHofer

### PR DESCRIPTION
**Pull request by ericstj**
Fix https://github.com/dotnet/sdk/issues/31655

### CODE REVIEW

**Review by ViktorHofer — APPROVED** [review]
Lovely. When we wrote this, we believed that a default without specifying the underlying type is sufficient. We were wrong.

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs [hunk 1]
```diff
@@ -1284,6 +1284,43 @@ public partial class C : B
...
+                        }
+
+                        public partial class B : A
+                        {
+                            public B() : base(default(char)) {}
+                        }
+                    }
+                    """);
+        }
+
```
Comment (carlossanlop) [inline src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1323]: @ericstj do you mind also adding test cases for the examples provided in the issue? https://github.com/dotnet/sdk/issues/31655

One default is casted to a class type, the other one to a generic type:

```cs
base((Packaging.Core.PackageIdentity)default!, (System.Collections.Generic.IEnumerable<Packaging.Core.PackageDependency>)default!)
```

Want to make sure both work as expected.

Comment (ericstj) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1323]: Sure I can, though I don't think we need to use those NuGet types.  I'll add some user types.

Comment (ericstj) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1323]: Interesting, so I'm adding the type arguments to default correctly, but it's being removed by https://github.com/dotnet/sdk/blob/4ea3e8304c21022bbea72a29154b7602af4f9239/src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs#L94

This seems to be a bug in the Reducer in roslyn.

Comment (ericstj) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1323]: Filed https://github.com/dotnet/roslyn/issues/67867 and I will add the regression test with a suppression.

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**